common/workunit/workunit.cpp

```diff
@@ -332,113 +332,1203 @@ extern WORKUNIT_API IConstWUGraphProgress *createConstGraphProgress(const char *
+    return compareScopeName(l->queryName(), r->queryName());
+}
+
+static int compareSubGraphNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    return l->getPropInt("@id") - r->getPropInt("@id");
+}
+
+static int compareActivityNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    return l->getPropInt("@id") - r->getPropInt("@id");
+}
+
+static int compareEdgeNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    //MORE: Edge needs more work
+    const char * leftId = l->queryProp("@id");
+    const char * rightId = r->queryProp("@id");
+    unsigned leftAc = atoi(leftId);
+    unsigned rightAc = atoi(rightId);
+    if (leftAc != rightAc)
+        return (int)(leftAc - rightAc);
+    const char * leftSep = strchr(leftId, '_');
+    const char * rightSep = strchr(rightId, '_');
+    assertex(leftSep && rightSep);
+    return atoi(leftSep+1) - atoi(rightSep+1);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+/*
+ * A class for implementing a scope iterator that walks through graph progress information
+ * Very similar to CConstGraphProgressStatisticsIterator.
+ * The code has been cloned rather than commoned up because CConstGraphProgressStatisticsIterator
+ * will be deprecated and deleted once this is fully functional (or implemented in terms of this).
+ */
+class CConstGraphProgressScopeIterator : public CInterfaceOf<IConstWUScopeIterator>
+{
+public:
+    CConstGraphProgressScopeIterator(const char * wuid, const IStatisticsFilter * _filter) : filter(_filter)
+    {
+        if (filter)
+            scopes.appendList(filter->queryScope(), ":");
+        const char * searchGraph = "*";
+        if (scopes.ordinality())
+            searchGraph = scopes.item(0);
+
+        rootPath.append("/GraphProgress/").append(wuid).append('/');
+        bool singleGraph = false;
+        if (!containsWildcard(searchGraph))
+        {
+            rootPath.append(searchGraph).append("/");
+            singleGraph = true;
+        }
+
+        //Don't lock the statistics while we iterate - any partial updates must not cause problems
+        if (daliClientActive())
+            conn.setown(querySDS().connect(rootPath.str(), myProcessSession(), RTM_NONE, SDS_LOCK_TIMEOUT));
+
+        if (conn && !singleGraph)
+        {
+            graphIter.setown(conn->queryRoot()->getElements("*"));
+            graphIter.setown(createSortedIterator(*graphIter, compareGraphNode));
+        }
+
+        valid = false;
+    }
+
+    IMPLEMENT_IINTERFACE_USING(CInterfaceOf<IConstWUScopeIterator>)
+
+    virtual bool first()
+    {
+        valid = false;
+        if (!conn)
```

Comment:
conn can be null if someone requested to filter by a non existent graph.  So I'm not sure it is sensible to distinguish them.